core/chains/chain_kv.go

```diff
@@ -0,0 +1,78 @@
+package chains
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+type ChainsKV[T ChainService] struct {
+	mu     sync.Mutex
+	chains map[string]T
+}
+
+var ErrNoSuchChainID = errors.New("chain id does not exist")
+
+func NewChainsKV[T ChainService]() *ChainsKV[T] {
+	return &ChainsKV[T]{
+		chains: make(map[string]T),
+	}
+}
+func (c *ChainsKV[T]) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.chains)
+}
+
+// Get return [ErrNoSuchChainID] if [id] is not found
+func (c *ChainsKV[T]) Get(id string) (T, error) {
+	var dflt T
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	chn, exist := c.chains[id]
+	if !exist {
+		return dflt, fmt.Errorf("%w: %s", ErrNoSuchChainID, id)
+	}
+	return chn, nil
+}
+
+func (c *ChainsKV[T]) Put(id string, chn T) {
```

Comment:
I'm surprised we still need mutability. Is `Put` used outside of tests and initialization?

Reply:
Is there any chance of simplifying to just:
```go
type ServiceMap[T ChainService] map[string]T
```
And then dropping `Get/Put/Slice`?

Reply:
Oh these are probably implementing interfaces - well could drop the lock anyways.

Reply:
yes, these are implementing interfaces to access LegacyChains.

Reply:
Even `Put`?